### PR TITLE
Crawl with bindings, not a whole bunch of AST transformers

### PR DESCRIPTION
**Pull request by jridgewell**

### CODE REVIEW

**Comment by jdalton** [comment]
\cc @thejameskyle for guidance.

File: src/Store.js [hunk 1]
```diff
@@ -25,39 +12,14 @@ export default class Store extends MapCache {
...
-  getMapBy(type, key) {
-    return _.invoke(this.getStoreBy(type, key), 'get', type);
-  }
-
-  getValueBy(type, key) {
-    return _.invoke(this.getMapBy(type, key), 'get', key);
-  }
-
-  set(pkgPath, pkgStore=new PackageStore(normalize(pkgPath))) {
+  set(pkgPath, pkgStore=new PackageStore(pkgPath)) {
```
Comment (jridgewell) [inline src/Store.js]: Need to normalize here too. You changed so many things last night. 😝

File: src/index.js [hunk 2]
```diff
@@ -29,17 +29,6 @@ export default function lodash({ types }) {
...
-    let { parent } = path;
-
-    // Trace curried calls to their origin, e.g. `fp.partial(func)([fp, 2])(1)`.
-    while (types.isCallExpression(parent)) {
-      parent = result = parent.callee;
-    }
-    return result;
-  }
-
   function isIdentifier(node, path) {
```
Comment (jridgewell) [inline src/index.js]: Can be dropped.

File: src/index.js [hunk 3]
```diff
@@ -52,185 +41,159 @@ export default function lodash({ types }) {
     return types.isImportDeclaration(node) && store.has(node.source.value);
```
Comment (jridgewell) [inline src/index.js]: This can be dropped.

**Comment by megawac** [comment]
What are the benefits of using bindings @jridgewell? Is this a perf win?

**Comment by jridgewell** [comment]
> What are the benefits of using bindings

For one, we'll never have [this problem](https://github.com/lodash/babel-plugin-lodash/issues/87) again. New AST type gets added? No problem, binding references will take care of it.

> Is this a perf win?

Huge. Imagine how many thousands of `ArrayExpression`s, `MemberExpression`s, `CallExpression`s, `AssignmentExpression`s, `ConditionalExpression`s, `ExpressionStatement`s, `JSXExpressionContainer`s, `LogicalExpression`s, `ObjectProperty`s, and `VariableDeclarator`s there are in even a normal app. Calling a function on every one of them is not scalable.

**Comment by megawac** [comment]
Is there a good reference on bindings? Is the section in the [babel-handbook](https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-bindings) the best one out there?

**Comment by jridgewell** [comment]
Yah, I usually refer to the babel-plugin-handbook for references. I actually find most of the internal concepts by playing around in a debugger on astexplorer.net.

File: src/Store.js (hunk 1 above)
Comment (jdalton) [inline src/Store.js]: You can rebase them in.

File: src/index.js (hunk 2 above)
Comment (jdalton) [inline src/index.js]: Is this a note for you to rebase or something I need to do?

File: src/index.js (hunk 3 above)
Comment (jdalton) [inline src/index.js]: Same [q](https://github.com/lodash/babel-plugin-lodash/pull/90/files#r70677343).

File: src/index.js [hunk 4]
```diff
@@ -52,185 +41,159 @@ export default function lodash({ types }) {
...
+        }
       });
+    });
 
-      // Clear tracked Lodash method imports and variables.
-      importModule.cache.clear();
-      store.clear();
-    },
+    // Phase 3: Remove the old imports.
+    _.each(declarations, (dec) => dec.remove());
```
Comment (jdalton) [inline src/index.js]:
```js
_.invokeMap(declarations, 'remove');
```

File: src/index.js [hunk 5]
```diff
@@ -52,185 +41,159 @@ export default function lodash({ types }) {
...
-          identifierMap.set(identifier.name, local.name);
-          memberMap.set(identifier.name, identifier);
-          memberMap.set(local.name, identifier);
-        }
-        else {
-          // Cache original default member name, e.g. `_` of `import _ from 'lodash'`.
-          defaultMap.set(local.name, true);
-          memberMap.set(local.name, identifiers.UNDEFINED);
+      this.declarations.push(path);
+      _.each(path.get("specifiers"), (spec) => {
```
Comment (jdalton) [inline src/index.js]: Single quotes.

File: src/index.js [hunk 6]
```diff
@@ -52,185 +41,159 @@ export default function lodash({ types }) {
...
-      else if (types.isBindExpression(callee)) {
-        visitor.CallExpression(path.get('callee'));
-      }
-      else if (types.isMemberExpression(callee)) {
-        visitor.MemberExpression(path.get('callee'));
-      }
-      // Replace Lodash references in arguments, e.g. `_.flow(_.map, _.head)`.
-      replaceArguments(node.callee, path);
+      const imports = {
+        defaults: [],
```
Comment (jdalton) [inline src/index.js]: *style nit*: Quote property names in object literals.

File: src/index.js (hunk 2 above)
Comment (jridgewell) [inline src/index.js]: Notes for me for when I get home.

File: src/index.js [hunk 7]
```diff
@@ -52,185 +41,159 @@ export default function lodash({ types }) {
...
+        } else {
+          const callee = getCallee(path);
+          if (callee && pkgStore.has(callee.name)) {
+            path.replaceWith(types.memberExpression(
+              callee,
+              types.identifier('placeholder')
+            ));
+          } else {
+            throw path.buildCodeFrameError('wtf');
+          }
```
Comment (jdalton) [inline src/index.js]: Use a more meaningful error message.

Comment (jridgewell) [inline src/index.js]: Lolz. It was late and I was getting frustrated with `[fp]` not transforming into `partial.placeholder`.

Comment (jdalton) [inline src/index.js]: It effects chaining error tests if that helps.

Comment (jridgewell) [inline src/index.js]: I managed to fix it (had to import everything before replacing any bindings), but I never updated my error message.

Comment (jridgewell) [inline src/index.js]: PS. Please comment with the error message you'd like. 😁

File: src/index.js [hunk 8]
```diff
@@ -52,185 +41,159 @@ export default function lodash({ types }) {
...
+
+        if (parentPath.isMemberExpression()) {
+          const name = path.parent.property.name;
+          parentPath.replaceWith(pkgStore.get(name));
+        } else {
+          const callee = getCallee(path);
+          if (callee && pkgStore.has(callee.name)) {
+            path.replaceWith(types.memberExpression(
+              callee,
+              types.identifier('placeholder')
```
Comment (jdalton) [inline src/index.js]: You can use `identifiers.PLACEHOLDER`.

File: src/index.js (hunk 7 above)
Comment (jdalton) [inline src/index.js]: No worries, I'm already prepping a clean up pass.

**Pull request merged**